[IMP] stock_storage_type: Allow to re-apply putaway rules on computed location for a selected move line

rousseldenis opened the pull request:
We must ensure the same destination is computed at each time.

During recomputation, this:

- Allows to keep the same destination if it is still allowed
- Allows to change the destination if it is not allowed anymore


- [x] Improve the writing of domains appending with `AND()`

rousseldenis committed:
[IMP] stock_storage_type: Allow to re-apply putaway rules on computed location for a selected move line

rousseldenis commented:
@jbaudoux @lmignon @phschmidt

lmignon commented on stock_storage_type/models/stock_storage_category_capacity.py:
```diff
@@ -95,7 +96,30 @@ def _domain_location_storage_type(self, candidate_locations, quants, products):
         # Build the domain using the 'allow_new_product' field
         if self.allow_new_product == "empty":
-            location_domain.append(("location_is_empty", "=", True))
+            # We should include the destination location of the current
+            # stock move line to avoid excluding it if already selected
```
@rousseldenis Can you improve your comment to describe the use case / context? It's hard to figure out the purpose of this code when reading the comment.

rousseldenis replied:
Improved. Thanks

rousseldenis committed:
[FIX] stock_storage_type: Allow to recompute putaways if location has been populated

In the case of no product mixing, if a destination location has already been computed and
in the meantime, the location has been populated by another product, the computation
still propose the same location.

Exclude locations in that case.

rousseldenis committed:
[IMP] stock_storage_type: Use of AND() instead of appending domains

rousseldenis committed:
[IMP] stock_storage_type: Rely on ORM in order to get locations per quantity

The current implementation was triggerring queries (read_group). Rely now
on ORM to get records as 'quant_ids' field exists on locations
